File: app/src/main/java/com/muravyovdmitr/notter/recyclers/NotesViewHolder.java
package com.muravyovdmitr.notter.recyclers;

import android.content.DialogInterface;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.muravyovdmitr.notter.ParamAction;
import com.muravyovdmitr.notter.models.Note;
import com.muravyovdmitr.notter.notter.R;


/**
 * Created by Dima Muravyov on 31.07.2016.
 */

class NotesViewHolder extends RecyclerView.ViewHolder implements Bindeable<Note> {
    // TODO: 31.07.2016 Injects this views
    private TextView title;
    private TextView message;

    private Removable removable;
    private ParamAction<Integer> onItemClick;
    ;

    NotesViewHolder(View itemView, Removable removable, ParamAction<Integer> onItemClick) {
        super(itemView);

        this.title = (TextView) itemView.findViewById(R.id.activity_note_item_title);
        this.message = (TextView) itemView.findViewById(R.id.activity_note_item_message);

        this.removable = removable;
        this.onItemClick = onItemClick;

        itemView.setOnClickListener(this.clickListener);
        itemView.setOnLongClickListener(this.longClickListener);
    }

    @Override
    public void bind(Note item) {
        this.title.setText(item.getTitle());
        this.message.setText(item.getMessage());
    }

    private AlertDialog.Builder removeItemDialog() {
        return new AlertDialog.Builder(this.itemView.getContext())
                .setTitle("Delete item #" + this.getAdapterPosition())
                .setMessage("Are you sure?")
                .setPositiveButton("Delete", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        removable.remove(getAdapterPosition());
                    }
                })
                .setNegativeButton(android.R.string.cancel, null);
    }

    private final View.OnClickListener clickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            onItemClick.execute(getAdapterPosition());
        }
    };

    private final View.OnLongClickListener longClickListener = new View.OnLongClickListener() {
        @Override
        public boolean onLongClick(View view) {
            removeItemDialog().show();
            return false;
        }
    };
}

File: app/src/main/java/com/muravyovdmitr/notter/recyclers/NotesViewHolder.kt
package com.muravyovdmitr.notter.recyclers

import android.support.v7.app.AlertDialog
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.TextView
import com.muravyovdmitr.notter.ParamAction
import com.muravyovdmitr.notter.models.Note
import com.muravyovdmitr.notter.notter.R


/**
 * Created by Dima Muravyov on 31.07.2016.
 */

public class NotesViewHolder(itemView: View, private val removable: Removable, private val onItemClick: ParamAction<Int>) : RecyclerView.ViewHolder(itemView), Bindeable<Note> {
    // TODO: 31.07.2016 Injects this views
    private val title: TextView
    private val message: TextView

    private val clickListener: View.OnClickListener
    private val longClickListener: View.OnLongClickListener

    init {
        this.title = itemView.findViewById(R.id.activity_note_item_title) as TextView
        this.message = itemView.findViewById(R.id.activity_note_item_message) as TextView

        this.clickListener = View.OnClickListener { onItemClick.execute(adapterPosition) }

        this.longClickListener = View.OnLongClickListener {
            removeItemDialog().show()
            false
        }

        itemView.setOnClickListener(this.clickListener)
        itemView.setOnLongClickListener(this.longClickListener)
    }

    override fun bind(item: Note) {
        this.title.text = item.title
        this.message.text = item.message
    }

    private fun removeItemDialog(): AlertDialog.Builder {
        return AlertDialog.Builder(this.itemView.context).setTitle("Delete item #" + this.adapterPosition).setMessage("Are you sure?").setPositiveButton("Delete") { dialogInterface, i -> removable.remove(adapterPosition) }.setNegativeButton(android.R.string.cancel, null)
    }
}
